torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -87,52 +87,95 @@ struct SymbolicShapeAnalyzer {
   c10::SymbolicShape run() {
     // TODO: only run while the last iteration has made a change
     for (size_t i = 0; i < 6; i++) {
-      substituteInputTensorProperties();
+      // XXX: we cannot substitute symbolic dims before passes like constant
+      // propagation, or we might inadvertently use them in arithmetic or
+      // other operators
+      substituteInputTensorProperties(/*substitute_symbolic_dims*/ false);
       LowerSimpleTuples(graph_);
       RemoveListMutation(graph_);
       UnrollConstantLoops(graph_);
       ConstantPropagation(graph_);
       PeepholeOptimize(graph_);
       ConstantPropagation(graph_);
+      EliminateCommonSubexpression(graph_);
     }
-    ConstantPooling(graph_);
-    EliminateDeadCode(graph_);
+    substituteInputTensorProperties(/*substitute_symbolic_dims*/ true);
+    // XXX: do not run any passes aftr we have substituted in symbolic dimension
+    // value, we do it so they can be easily extracted into the output shape
     return extractOutputShape();
   }
 
  private:
-  void substituteInputTensorProperties() {
-    for (auto index : tensor_indices) {
-      substituteTensorProperties(index);
+  void substituteInputTensorProperties(bool substitute_symbolic_dims) {
+    std::unordered_map<int64_t, std::vector<Value*>> symbolic_shape_map;
```

Comment:
Im having hard times following what's going on here: do you mind adding some comments, maybe with examples?

Reply:
Note that readers of that code would not first see the commit message or the test coming in this PR - it's good for the code to have a similar description too.